python/ucxx/_lib/libucxx.pyx

```diff
@@ -40,15 +41,45 @@ include "tag.pyx"
 logger = logging.getLogger("ucx")
 
 
-np.import_array()
-
+cdef class HostBufferAdapter:
+    """A simple adapter around HostBuffer implementing the buffer protocol"""
+    cdef Py_ssize_t _size
+    cdef void* _ptr
+    cdef Py_ssize_t[1] _shape
+    cdef Py_ssize_t[1] _strides
+    cdef Py_ssize_t _itemsize
+
+    @staticmethod
+    cdef _from_host_buffer(HostBuffer* host_buffer):
+        """Construct a new HostBufferAdapter from a HostBuffer.
+
+        This factory takes ownership of the input host_buffer's data, so
+        attempting to use the input after this function is called will result
+        in undefined behavior.
+        """
+        cdef HostBufferAdapter obj = HostBufferAdapter.__new__(HostBufferAdapter)
+        obj._size = host_buffer.getSize()
+        obj._ptr = host_buffer.release()
+        obj._shape = [obj._size]
+        obj._itemsize = sizeof(uint8_t)
+        obj._strides = [obj._itemsize]
+        return obj
+
+    def __getbuffer__(self, Py_buffer *buffer, int flags):
+        buffer.buf = self._ptr
+        buffer.format = 'c'
+        buffer.internal = NULL
+        buffer.itemsize = self._itemsize
+        buffer.len = self._size * self._itemsize
+        buffer.ndim = 1
+        buffer.obj = self
+        buffer.readonly = 0
+        buffer.shape = self._shape
+        buffer.strides = self._strides
+        buffer.suboffsets = NULL
 
-cdef ptr_to_ndarray(void* ptr, np.npy_intp N):
-    cdef np.ndarray[np.uint8_t, ndim=1, mode="c"] arr = (
-        np.PyArray_SimpleNewFromData(1, &N, np.NPY_UINT8, <np.uint8_t*>ptr)
-    )
-    PyArray_ENABLEFLAGS(arr, NPY_ARRAY_OWNDATA)
-    return arr
+    def __releasebuffer__(self, Py_buffer *buffer):
+        free(self._ptr)
```

Comment:
Good call on tying the lifetime to this object instead, I agree that makes more sense.

> you could still implement it, since it disallows some limited downstream usage so if we need it later having it blocks that, but it doesn't matter really

Could you explain this statement? For now I've changed the implementation to a no-op, but it sounds like you're saying the mere existence of this method changes the behavior in some potential meaningful (and potentially desirable?) way.

Reply:
CPython has some the parsing functions `PyArg_ParseTuple("%s", buffer)` which only work if a buffer-like doesn't implement `__releasebuffer__` (because that usage pattern cannot call the release buffer function).
So implementing it as a no-op disallows those contexts and may make adding it later easier.

I noticed that cupy also does that for it's pinned memory, but not sure they had a specific reason.

Reply:
I see, I've seen those functions but I didn't realize that the existence of this method would block that. Good context, thanks.